File: targets/java/src/main/java/io/crowbar/diagnosis/algorithms/DiagnosticSystem.java
package io.crowbar.diagnosis.algorithms;

import io.crowbar.diagnosis.spectra.Spectra;

import flexjson.JSON;
import flexjson.JSONSerializer;
import flexjson.transformer.AbstractTransformer;

import java.util.ArrayList;
import java.util.List;
import io.crowbar.diagnosis.spectra.EditableSpectra;
import io.crowbar.diagnosis.spectra.serializers.HitSpectraSerializer;

public final class DiagnosticSystem {
    private static class ExcludeTransformer extends AbstractTransformer {
        @Override
        public Boolean isInline () {return true;}

        @Override
        public void transform (Object object) {}
    }

    private static class Connection {
        private int from;
        private int to;
        Connection (int from, int to) {
            this.from = from;
            this.to = to;
        }

        private int getFrom () {
            return from;
        }

        private int getTo () {
            return to;
        }
    }

    private final List<Algorithm> generators = new ArrayList<Algorithm> ();
    private final List<Algorithm> rankers = new ArrayList<Algorithm> ();
    private final List<Connection> connections = new ArrayList<Connection> ();

    // ! Adds a generator to the diagnostic system. (Note: makes a copy of the generator so its safe to reuse it.)
    public int addGenerator (Generator g) {
        generators.add(new Algorithm(g.getAlgorithm()));
        return generators.size() - 1;
    }

    // ! Adds a ranker to the diagnostic system. (Note: makes a copy of the ranker so its safe to reuse it.)
    public int addRanker (Ranker r) {
        rankers.add(new Algorithm(r.getAlgorithm()));
        return rankers.size() - 1;
    }

    public void addConnection (int generatorId,
                               int rankerId) {
        assert generatorId >= 0 && generatorId < generators.size();
        assert rankerId >= 0 && rankerId < rankers.size();
        connections.add(new Connection(generatorId, rankerId));
    }

    @JSON
    private List<Algorithm> getGenerators () {
        return generators;
    }

    @JSON
    private List<Algorithm> getRankers () {
        return rankers;
    }

    @JSON
    private List<Connection> getConnections () {
        return connections;
    }

    public void run (Spectra spectra) {
        String ret = this.run(HitSpectraSerializer
                              .serialize(spectra),
                              this.toString());
    }

    public native String run (String spectra, String configs);


    @Override
    public String toString () {
        return new JSONSerializer()
               .transform(new ExcludeTransformer(), void.class)
               .exclude("*.class")
               .deepSerialize(this);
    }

    public static void main (String[] args) {
        DiagnosticSystem j = new DiagnosticSystem();


        j.addGenerator(new SingleFaultGenerator());
        j.addRanker(new SimilarityRanker(SimilarityRanker.Type.OCHIAI));
        j.addConnection(0, 0);

        j.addGenerator(new MHSGenerator());
        j.addRanker(new FuzzinelRanker());
        j.addConnection(0, 1);
        j.addConnection(1, 1);

        // j.run(new EditableSpectra());
        System.out.println("DiagnosticSystem: " + j);
    }
}

File: diagnostic-native/libdiag/diagnostic_system.cpp
#include "diagnostic_system.h"

#include "utils/iostream.h"

namespace diagnostic {

void t_diagnostic_system::add_generator (t_const_ptr<t_candidate_generator> generator) {
    generators.push_back(generator);
}

const t_diagnostic_system::t_generators & t_diagnostic_system::get_generators () const {
    return generators;
}

void t_diagnostic_system::add_ranker (t_const_ptr<t_candidate_ranker> ranker) {
    rankers.push_back(ranker);
}

const t_diagnostic_system::t_rankers & t_diagnostic_system::get_rankers () const {
    return rankers;
}

void t_diagnostic_system::add_connection (const t_connection & c) {
    assert(c.get_from() >= 0);
    assert(c.get_from() < generators.size());
    assert(c.get_to() >= 0);
    assert(c.get_to() < rankers.size());

    connections.push_back(c);
}

void t_diagnostic_system::add_connection (t_id generator_id, t_id ranker_id) {
    add_connection(t_connection(generator_id, ranker_id));

}

const t_diagnostic_system::t_connections & t_diagnostic_system::get_connections () const {
    return connections;
}
}

namespace std {

std::ostream& operator<<(std::ostream & s,
                         const diagnostic::t_candidate_generator& cg) {
    s << cg.to_string();
    return s;
}

std::ostream& operator<<(std::ostream & s,
                         const diagnostic::t_candidate_ranker & cr) {
    s << cr.to_string();
    return s;
}

std::ostream& operator<<(std::ostream& s,
                         const diagnostic::t_diagnostic_system & ds){
    s << "t_diagnostic_system[generators:" << ds.get_generators() << ", ";
    s << "rankers:" << ds.get_rankers() << ", ";
    s << "connections:" << ds.get_connections() << "]";
    return s;
}
}
